src/components/OnboardingWelcomeVideo.tsx

```diff
@@ -0,0 +1,166 @@
+
+// Aspect ratio and height of the video.
+// Useful before video loads to reserve space.
+const VIDEO_ASPECT_RATIO = 1280 / 960;
+
+const MODAL_PADDING = variables.spacing2;
+
+type VideoLoadedEventType = {
+    srcElement: {
+        videoWidth: number;
+        videoHeight: number;
+    };
+};
+
+type VideoPlaybackStatusEventType = {
+    isLoaded: boolean;
+};
+
+type VideoStatus = 'video' | 'animation';
+
+function OnboardingWelcomeVideo() {
+    const {translate} = useLocalize();
+    const styles = useThemeStyles();
+    const [isModalVisible, setIsModalVisible] = useState(true);
+    const {shouldUseNarrowLayout} = useOnboardingLayout();
+    const [welcomeVideoStatus, setWelcomeVideoStatus] = useState<VideoStatus>('video');
+    const [isWelcomeVideoStatusLocked, setIsWelcomeVideoStatusLocked] = useState(false);
+    const [isVideoLoaded, setIsVideoLoaded] = useState(false);
+    const [videoAspectRatio, setVideoAspectRatio] = useState(VIDEO_ASPECT_RATIO);
+    const {isOffline} = useNetwork();
+
+    useEffect(() => {
+        if (isWelcomeVideoStatusLocked) {
+            return;
+        }
+
+        if (isOffline) {
+            setWelcomeVideoStatus('animation');
+        } else if (!isOffline && isVideoLoaded) {
+            setWelcomeVideoStatus('video');
+            setIsWelcomeVideoStatusLocked(true);
+        }
+    }, [isOffline, isVideoLoaded, isWelcomeVideoStatusLocked]);
+
+    const closeModal = useCallback(() => {
+        setIsModalVisible(false);
+        Navigation.goBack();
+    }, []);
+
+    const setAspectRatio = (event: VideoReadyForDisplayEvent | VideoLoadedEventType | undefined) => {
+        if (!event) {
+            return;
+        }
+
+        if ('naturalSize' in event) {
+            setVideoAspectRatio(event.naturalSize.width / event.naturalSize.height);
+        } else {
+            setVideoAspectRatio(event.srcElement.videoWidth / event.srcElement.videoHeight);
+        }
+    };
+
+    const setVideoStatus = (e: VideoPlaybackStatusEventType) => {
+        setIsVideoLoaded(e.isLoaded);
+    };
+
+    const getWelcomeVideo = () => {
+        const aspectRatio = videoAspectRatio || VIDEO_ASPECT_RATIO;
+
+        return (
+            <View
+                style={[
+                    styles.w100,
+                    // Prevent layout jumps by reserving height
+                    // for the video until it loads. Also, when
+                    // welcomeVideoStatus === 'animation' it will
+                    // set the same aspect ratio as the video would.
+                    {aspectRatio},
+                ]}
+            >
+                {welcomeVideoStatus === 'video' ? (
+                    <VideoPlayer
+                        url={CONST.WELCOME_VIDEO_URL}
+                        videoPlayerStyle={[styles.onboardingVideoPlayer, {aspectRatio}]}
+                        onVideoLoaded={setAspectRatio}
+                        onPlaybackStatusUpdate={setVideoStatus}
```

Comment:
Hmm not sure if we can. `onVideoLoaded` doesn't have access to `isLoaded` property. Let me just add a `useCallback` to `setVideoStatus`. The performance hit isn't big anyway since it's using `setIsVideoLoaded(event.isLoaded);` and the value of `isLoaded` stays always true. The invocations are probably due to `onPlaybackStatusUpdate` being called each time the progress bar of the video gets updated.

<img width="629" alt="image" src="https://github.com/Expensify/App/assets/85628901/f3a4acfa-b49d-4a73-a9fe-6bf0d62d9a4b">

Reply:
From `BaseVideoPlayer.js`:
```
  /**
   * A function to be called regularly with the `AVPlaybackStatus` of the video. You will likely be using this a lot.
   * See the [AV documentation](./av) for further information on `onPlaybackStatusUpdate`, and the interval at which it is called.
   * @param status
   */
  onPlaybackStatusUpdate?: (status: AVPlaybackStatus) => void;
/**
   * A function to be called when the video is ready for display. Note that this function gets called whenever the video's natural size changes.
   * @param event
   */
  onReadyForDisplay?: (event: VideoReadyForDisplayEvent) => void; <- this is what becomes onVideoLoaded
```

Reply:
Let me try to play around with it some more

Reply:
Seems like we need that `onPlaybackStatusUpdate` callback. Without it, when I tried to move `setVideoStatus` to `onVideoLoaded`, video would not start playing when going from offline into online. as only `onPlaybackStatusUpdate` would call once video loads after going back online.